### PR TITLE
adding the enable_gpu_compatible_memory pass

### PR DESCRIPTION
**Pull request by kushanam**
The purpose for this PR is to add a data grappler pass to enable GPU compatible mode for tensors which will be staged downstream using prefetch_to_device. This is in accordance with recent efforts to parallelize the fetch and execute (commit #s: 9a8772d, ad60111, 2803dfc, 4111779, ecc3db7).
Note that the purpose for this PR would be to enable the pass. There would be a subsequent PR to implement the "UseGpuAllocator" potentially within the following data ops: RepeatDataset, MapDataset, ParallelMapDataset, ParallelMapDatasetV2, InterleaveDataset, ParallelInterleaveDatasetV2, ParallelInterleaveDatasetV3, ParallelInterleaveDatasetV4, FilterDataset, ParallelFilterDataset, BatchDataset, BatchDatasetV2, ParallelBatchDataset, ShardDataset, PaddedBatchDataset, PaddedBatchDatasetV2, RangeDataset
Attn: @changhuilin

### CODE REVIEW

**Comment by gbaned** [comment]
@qqfish Can you please review this PR ? Thank you!

**Comment by kushanam** [comment]
@reedwm, @gbaned  do you know what is the  "assign-partner-tags / assign-partner-tags" about? It was passing before and seems like rerunning the pipeline may mitigate the issue. If so, could you rerun it please?

**Comment by reedwm** [comment]
>  @reedwm, @gbaned do you know what is the "assign-partner-tags / assign-partner-tags" about? It was passing before and seems like rerunning the pipeline may mitigate the issue. If so, could you rerun it please?

/CC @mihaimaruseac, do you know?

**Comment by mihaimaruseac** [comment]
It shouldn't be blocking. It's part of our experiment to give broader access to certain contributors.

**Comment by nluehr** [comment]
Any update @qqfish?

File: tensorflow/core/grappler/optimizers/data/enable_gpu_compatible_memory.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    if (!HasUseGpuCompatAllocatorAttr(*node_prior))
+    {
+      VLOG(2) << "The `" << node_prior->op() << "` op does not have the "
+        << UseGpuCompatAllocatorAttr << " attribute, hence the "
+        << "OptimizeAndCollectStats is skipped";
+      return Status::OK();
+    }
+    (*node_prior->mutable_attr())[UseGpuCompatAllocatorAttr].set_b(true);
```
Comment (qqfish) [inline tensorflow/core/grappler/optimizers/data/enable_gpu_compatible_memory.cc:70]: Should we ensure the next op is placed on GPU before adding this attribute?

File: tensorflow/core/framework/dataset_options.proto [hunk 2]
```diff
@@ -144,6 +144,12 @@ message OptimizationOptions {
   oneof optional_filter_parallelization {
     bool filter_parallelization = 18;
   }
+
+  // Whether to use gpu compatible memory prior to the 
+  // prefetch_to_device call
+  oneof optional_enable_gpu_compatible_memory {
+    bool enable_gpu_compatible_memory = 19;
```
Comment (qqfish) [inline tensorflow/core/framework/dataset_options.proto:151]: I think today we can already overlapping CPU->GPU transfer with GPU computation after changhuilin@'s change, right? So is the goal here to save one memcpy on CPU?

Comment (kushanam) [reply tensorflow/core/framework/dataset_options.proto:151]: Yes, and in addition it is a step towards using the pinned allocator for outputs that will get copied to the device, so it is more robust and more efficient.

File: tensorflow/core/grappler/optimizers/data/enable_gpu_compatible_memory.cc (hunk 1 above)
Comment (kushanam) [reply tensorflow/core/grappler/optimizers/data/enable_gpu_compatible_memory.cc:70]: The next op is prefetch_to_device, so I think it is effectively on GPU,

Comment (qqfish) [reply tensorflow/core/grappler/optimizers/data/enable_gpu_compatible_memory.cc:70]: We use prefetch_to_device for TPU as well. Maybe check the device type?

**Comment by gbaned** [comment]
@kushanam Can you please resolve conflicts? Thank you!

**Comment by qqfish** [comment]
Can you resolve my last comment? Other than that it looks good.

**Comment by kushanam** [comment]
> @kushanam Can you please check @qqfish's comments and keep us posted ? Thank you!

The last comment I got was regarding the TPU check which is resolved. @qqfish if there is anything I am  missing, could you kindly point out?

**Review by qqfish — APPROVED** [review]

**Comment by ezhulenev** [comment]
@jsimsa Can you please share what are your concerns from a `tf.data` team perspective.

**Comment by jsimsa** [comment]
This PR alludes to the fact it might need to change the signature of a large number of tf.data TF ops. In addition, the changes to the options.proto also suggests there will be a corresponding change to the Python tf.data.Options API to se that option. 

The concern is that the tf.data team has no prior context of this change. In general, the team has a very high bar for extending signature of existing ops (e.g. because of FW and BW compatibility considerations). The fact that this PR suggests a large list of ops that will need to be modified creates a concern that the approach will not work custom tf.data (i.e. ops outside of tf.data repo) ops (unless those ops are updated accordingly). If so, this does not seem like a scalable solution and I wonder what alternatives (if any) have been considered.

For the above reasons, the tf.data team would like to review and approve the design proposal that provides justification for the above changes, weighing the pros and cons of different alternatives.

**Comment by gbaned** [comment]
@ezhulenev Any update on this PR? Please. Thank you!

**Comment by kushanam** [comment]
Closing this pull request. Currently working on an alternative solution with directions from @jsimsa

**Pull request closed without merging**